pageserver/src/tenant/mgr.rs

```diff
@@ -861,22 +865,58 @@ impl TenantManager {
                         // take our fast path and just provide the updated configuration
                         // to the tenant.
                         tenant.set_new_location_config(AttachedTenantConf::try_from(
-                            new_location_config,
+                            new_location_config.clone(),
                         )?);
 
-                        // Persist the new config in the background, to avoid holding up any
-                        // locks while we do so.
-                        // TODO
-
-                        return Ok(());
+                        Some(tenant.clone())
                     } else {
                         // Different generations, fall through to general case
+                        None
                     }
                 }
                 _ => {
                     // Not an Attached->Attached transition, fall through to general case
+                    None
+                }
+            }
+        };
+
+        // Fast-path continued: having dropped out of the self.tenants lock, do the async
+        // phase of waiting for flush, before returning.
+        if let Some(tenant) = modify_tenant {
+            // Transition to AttachedStale means we may well hold a valid generation
+            // still, and have been requested to go stale as part of a migration.  If
+            // the caller set `flush`, then flush to remote storage.
+            if let LocationMode::Attached(AttachedLocationConfig {
+                generation: _,
+                attach_mode: AttachmentMode::Stale,
+            }) = &new_location_config.mode
+            {
+                if let Some(flush_timeout) = flush {
+                    info!(timeout_ms = flush_timeout.as_millis(), "Flushing");
+                    match tokio::time::timeout(
+                        flush_timeout,
+                        tenant.flush_remote().instrument(info_span!("flush_remote")),
+                    )
+                    .await
+                    {
+                        Ok(r) => {
+                            if let Err(e) = r {
+                                tracing::error!("Failed to flush to remote storage: {e}");
+                                return Err(e);
+                            }
+                        }
```

Comment:
```suggestion
                        Ok(Ok(()) => {},
                        Ok(Err(e)) => {
                            tracing::error!("Failed to flush to remote storage: {e}");
                            return Err(e);
                        }
```

Why are we though logging and propagating the error?

Reply:
The `{e}` is probably wrong as well.